Custom Clustering: Integrate Cluster View into Expression Preview

zyadtaha opened the pull request:
This is the fourth PR in the GSoC project.

Changes proposed in this pull request:
- Changed the "Preview" label to "Expression Preview".
- Added a "Clusters Preview" tab.
- Increased the dialog height by 50px to avoid too much scrolling.
- ![Screenshot from 2024-07-26 11-23-42](https://github.com/user-attachments/assets/60cdf83a-41fa-4424-a913-023112c5c554)

zyadtaha commented:
I have noticed that the clusters rendered in the tab are currently static and do not change when the expression changes.
I will start making them dynamic based on the expression.

ostephens commented:
> I have noticed that the clusters rendered in the tab are currently static and do not change when the expression changes.
I will start making them dynamic based on the expression.

I don't know if it is an issue in this situation, but in the main clustering dialogue we have recently introduced an option to either "auto update cluster" or to wait until the user clicks a button 

<img width="961" alt="Screenshot 2024-07-26 at 10 57 18" src="https://github.com/user-attachments/assets/3f4d139a-fd90-4235-b452-c4ed09320821">

The reason for this is that in some situations auto-calculating the clusters as the user makes changes can be resource intensive and take time.

I'm not sure if the same concerns apply here but it may be worth considering whether auto-updating could cause performance issues in some situations

zyadtaha commented:
> The reason for this is that in some situations auto-calculating the clusters as the user makes changes can be resource intensive and take time.

@wetneb has addressed this issue and suggested having the user refresh the previews manually, as you mentioned. I think adding a button to update the clusters is a good idea. What are your thoughts?

ostephens commented:
>@wetneb has addressed this issue and suggested having the user refresh the previews manually, as you mentioned. I think adding a button to update the clusters is a good idea. What are your thoughts?

I'm happy either way to be honest. I'd start with the simplest implementation and then review if time allows

wetneb reviewed (commented):
Some usability comments after trying it out:
* Editing an existing expression (in the "Manage custom clustering functions" dialog) still brings the old dialog, without clustering preview
* Clicking on the "Cancel" or "OK" button in the clustering expression editor redirects me to "http://localhost:3333/project?" which does not load properly
* When the expression does not generate any clusters, the clusters preview is left empty
* When the expression supposedly generate clusters (such a constant expression `"foo"` for which all values should be clustered together), the clusters preview is still empty

![image](https://github.com/user-attachments/assets/04dd01fe-5e75-4114-96cb-f769c928e36d)

Concerning the code, I see you have duplicated the expression preview dialog to add the previews rendering. Ideally we should minimize the code duplication, but I haven't investigated if/how that would be possible. Is it something you considered?

zyadtaha commented:
Hi @wetneb, 
I've sent you an email, please check it.

I was dealing with this problem before participating in the ICPC competition, and I still haven't figured it out.
![Screenshot from 2024-07-31 22-04-05 (2)](https://github.com/user-attachments/assets/cadaa986-cc83-42fa-ae1c-5fee4203af16)

When I debug, it keeps going through the distance path, even though the function parameter is set to "UserDefinedKeyer." Do you know why this is happening?

Here's the request sent:

```
      clusterer: JSON.stringify({
          'type' : this._method,
          'function' : "UserDefinedKeyer",
          'column' : this._columnName,
          'params': {
              'expression': "value + '100'"
          }
      })
```

I’ve restarted IntelliJ and run the following commands multiple times, but nothing has changed:
```
./refine clean
./refine build
```

wetneb commented:
@zyadtaha that's an oddity of Java:
https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java

zyadtaha commented:
> * Clicking on the "Cancel" or "OK" button in the clustering expression editor redirects me to "http://localhost:3333/project?" which does not load properly

Could you give me more details? For instance, what do you see in the Console? @wetneb 
We can discuss the other points in our meeting today.

zyadtaha commented:
Hi @wetneb, This PR is ready to review.
Here's a demo. 

https://github.com/user-attachments/assets/a5a63d16-1bf7-40c7-b956-6ca2cbffb9dc

wetneb commented:
The video looks great! Sorry that I might take a bit longer to review this as I am at Wikimania

wetneb commented on main/webapp/modules/core/scripts/dialogs/expression-preview-dialog.js:
```diff
@@ -482,3 +524,62 @@ ExpressionPreviewDialog.Widget.prototype._renderPreview = function(expression, d
+        };
+
+        var maxRenderRows = parseInt(
+            Refine.getPreference("ui.clustering.choices.limit", 5000)
```
Because it's a preview it might make sense to use a lower cap (perhaps not configurable?). But it would mostly be useful if the DOM rendering is a bottleneck, which I am not sure of.

zyadtaha replied:
I used the debugger to test the Cluster preview with a distance expression set to `value1.length() - value2.length()` on this [dataset](https://github.com/user-attachments/files/16568521/lieux-de-tournage-a-paris.csv). It generated over 5000 clustering choices and the clusters were displayed quickly without any slowdowns.

![](https://github.com/user-attachments/assets/16df5d42-20ed-4b38-9d73-d1ea3f91dc74)

wetneb reviewed (commented):
Very nice, it's exciting to see things coming into place :)
With your video I realize that beyond the preview of the clusters themselves, it could be useful to have some overall statistics such as the total number of clusters (compared to the total number of rows) so that the user can quickly get a sense of how "collapsing" their expression is. I also wonder about the parameters for KNN clustering, see below.

zyadtaha commented:
> it could be useful to have some overall statistics such as the total number of clusters (compared to the total number of rows) so that the user can quickly get a sense of how "collapsing" their expression is.

How does this placement look? @wetneb 
![Screenshot from 2024-08-10 03-18-32](https://github.com/user-attachments/assets/6ae6f7b1-b6cf-4b0f-9a54-8c71c2c3a70e)

zyadtaha commented:
Another point to discuss is the preview for the distance expression, which isn’t very helpful right now. Should we keep it as is, or show a preview for expressions set to "value," for example?

![](https://github.com/user-attachments/assets/56dfe76a-2fed-43d7-827f-676066f1c4b4)

zyadtaha commented on main/webapp/modules/core/scripts/dialogs/expression-preview-dialog.js:
```diff
@@ -431,6 +439,40 @@ ExpressionPreviewDialog.Widget.prototype.update = function() {
+                    'function' : activeTabName === "Keying" ? "UserDefinedKeyer" : "UserDefinedDistance",
+                    'column' : self._columnName,
+                    'params' : {
+                        "expression" : expression
```
It already has default parameters: Radius is 1 and Block chars is 6 (same as the clustering dialog).

We have three options:
- Keep these defaults and make them clear to the user.
- Pass the parameters from the clustering dialog to our custom dialog.
- Add a small row in the preview so users can adjust these parameters.

I think adding a small row for user input is the best option. What do you think?

wetneb commented:
> How does this placement look? @wetneb

Looks sensible. Perhaps you could make the text a bit shorter so that it fits better? Something like "1234 rows, 567 clusters"?

> Another point to discuss is the preview for the distance expression, which isn’t very helpful right now. Should we keep it as is, or show a preview for expressions set to "value," for example?

Ah yes, I forgot that we haven't solved this problem yet!
I think the preview should be done in the same sort of context as the evaluation for clustering is done, so with `value1` and `value2` pointing to sample values from the column. It's perhaps not so easy to select interesting pairs of values automatically from the column though. This calls for a different design, I would say.

Instead of displaying the first few rows, we could have two input fields where the user would be able to set the `value1` and `value2` themselves. Those fields could be prefilled with the first two values we find in the column. It would also show the result of the evaluation on those two values, which would need to be a number (showing a warning if not).

I would say this would have the benefit of:
* making clear that `value1` and `value2` are available in the evaluation context
* showing the user what sort of values they are expected to hold
* making sure the expression is returning a number

But that's just a suggestion on top of my mind - it's likely worth suggesting a design on the forum.

wetneb commented:
I also noticed that some exceptions stack traces are shown in the logs:
```
08:47:44.856 [                   refine] POST /command/core/compute-clusters (6ms)
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: The string cannot be converted to a double
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at edu.mit.simile.vicino.clustering.NGramClusterer.getClustersMultiThread(NGramClusterer.java:138)
	at edu.mit.simile.vicino.clustering.NGramClusterer.getClusters(NGramClusterer.java:113)
	at com.google.refine.clustering.knn.kNNClusterer$BlockingClusteringRowVisitor.getClusters(kNNClusterer.java:227)
	at com.google.refine.clustering.knn.kNNClusterer.computeClusters(kNNClusterer.java:244)
	at com.google.refine.commands.browsing.ComputeClustersCommand.doPost(ComputeClustersCommand.java:91)
	at com.google.refine.RefineServlet.service(RefineServlet.java:187)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1410)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:578)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1570)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:790)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1384)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1543)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1306)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at com.google.refine.ValidateHostHandler.handle(ValidateHostHandler.java:93)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalArgumentException: The string cannot be converted to a double
	at com.google.refine.clustering.knn.UserDefinedDistance.convertToDouble(UserDefinedDistance.java:84)
	at com.google.refine.clustering.knn.UserDefinedDistance.compute(UserDefinedDistance.java:70)
	at com.google.refine.clustering.knn.kNNClusterer$BlockingClusteringRowVisitor$DistanceWrapper.d(kNNClusterer.java:192)
	at edu.mit.simile.vicino.clustering.NGramClusterer$BlockEvaluator.call(NGramClusterer.java:78)
	at edu.mit.simile.vicino.clustering.NGramClusterer$BlockEvaluator.call(NGramClusterer.java:51)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 more
Caused by: java.lang.NumberFormatException: For input string: "Coco and Co Prod"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:651)
	at com.google.refine.clustering.knn.UserDefinedDistance.convertToDouble(UserDefinedDistance.java:82)
```

It'd be cleaner not to have them, and report any errors that are worth reporting to the user (such as: GREL function not returning the expected datatype).

zyadtaha commented:
> Instead of displaying the first few rows, we could have two input fields where the user would be able to set the `value1` and `value2` themselves. Those fields could be prefilled with the first two values we find in the column. It would also show the result of the evaluation on those two values, which would need to be a number (showing a warning if not).

I made a design and posted it on the forum - please check it out [here](https://forum.openrefine.org/t/user-defined-clustering-project/1528/36)

zyadtaha commented:
### Recent changes made:

- Redesigned the interface so users can now input their own example values, with two fields that start with the same initial value (from the first value in the column) and dynamically preview the expression's result.
- Fixed the issue where exception stack traces were appearing in the logs—please verify this.
- Added an error message that shows up if the distance expression doesn't return a number.
- Included a small row in the preview where users can adjust the radius and blocking characters, enabling dynamic clustering based on these settings.

### Here's a demo:

[Screencast from 08-21-2024 03:30:20 PM.webm](https://github.com/user-attachments/assets/db3dc0b2-8767-4e53-9651-e4ef9b55e5e0)

wetneb commented:
Could it be just some non-determinism in the order of the returned clusters? Meaning that the clusters are still the same, just not returned in the same order? If that's indeed a problem, then maybe we can sort them with some criterion.

zyadtaha commented:
> Could it be just some non-determinism in the order of the returned clusters? Meaning that the clusters are still the same, just not returned in the same order? If that's indeed a problem, then maybe we can sort them with some criterion.

If you focus on the number of clusters in the video, you will see them varying. I also manually checked some clustering results, and they are not the same

zyadtaha commented:
> And the clusters preview tab shows "No clusters returned" instead of showing an error. Ideally, it would be better if we could have an error message such as "Computing the clusters failed", "The distance function returned a non-numeric value for inputs `"hello"` and `"world"`"

What do `"hello"` and `"world"`" represent? Shouldn't we only show `"Computing the clusters failed"`, `"The distance function returns a non-numeric value"`?

wetneb commented:
> What do "hello" and "world"" represent? Shouldn't we only show "Computing the clusters failed", "The distance function returns a non-numeric value"?

For some GREL expressions, a number will be returned in some cases, but not in others (for instance, it could return an error). In such cases, it is helpful to know for which arguments the GREL function did not return a number, so that it can be fixed appropriately. In my message above, the `"hello"` and `"world"` values represented a pair of sample values for which the GREL expression would not return a number. I thought it could help the user to have them included in the error message, so that they can understand what is going on more easily. But that's just a suggestion.

zyadtaha commented:
> > What do "hello" and "world"" represent? Shouldn't we only show "Computing the clusters failed", "The distance function returns a non-numeric value"?
> 
> For some GREL expressions, a number will be returned in some cases, but not in others (for instance, it could return an error). In such cases, it is helpful to know for which arguments the GREL function did not return a number, so that it can be fixed appropriately. In my message above, the `"hello"` and `"world"` values represented a pair of sample values for which the GREL expression would not return a number. I thought it could help the user to have them included in the error message, so that they can understand what is going on more easily. But that's just a suggestion.

Oh, I see now! That's actually a very good suggestion. We should definitely implement it.

However, I'm curious—why are we considering creating a new exception? I think we could clear the stack trace by adding a try-catch block to the class invoking the "compute" function (rather than the command class). We can also pass the pair of sample values in the exception message.

wetneb commented:
It's generally a good practice to have quite specific exception types, to make them more informative. Adding the sample values as attributes on this exception would let you serialize them separately in the JSON response sent to the frontend, enabling the frontend to render thos separately (for instance with a different font) to distinguish them from the rest of the error message.

zyadtaha commented:
```
java.util.concurrent.ExecutionException: java.lang.RuntimeException: com.google.refine.clustering.knn.DistanceCalculationFailedException: Computing the clusters failed, The distance function returned a non-numeric value for inputs THALIE IMAGES and THALIE IMAGES - TTMT
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at edu.mit.simile.vicino.clustering.NGramClusterer.getClustersMultiThread(NGramClusterer.java:138)
	at edu.mit.simile.vicino.clustering.NGramClusterer.getClusters(NGramClusterer.java:113)
	at com.google.refine.clustering.knn.kNNClusterer$BlockingClusteringRowVisitor.getClusters(kNNClusterer.java:231)
	at com.google.refine.clustering.knn.kNNClusterer.computeClusters(kNNClusterer.java:248)
	at com.google.refine.commands.browsing.ComputeClustersCommand.doPost(ComputeClustersCommand.java:117)
	at com.google.refine.RefineServlet.service(RefineServlet.java:187)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1410)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:578)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1570)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:790)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1384)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1543)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1306)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at com.google.refine.ValidateHostHandler.handle(ValidateHostHandler.java:93)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: com.google.refine.clustering.knn.DistanceCalculationFailedException: Computing the clusters failed, The distance function returned a non-numeric value for inputs THALIE IMAGES and THALIE IMAGES - TTMT
	at com.google.refine.clustering.knn.kNNClusterer$BlockingClusteringRowVisitor$DistanceWrapper.d(kNNClusterer.java:195)
	at edu.mit.simile.vicino.clustering.NGramClusterer$BlockEvaluator.call(NGramClusterer.java:78)
	at edu.mit.simile.vicino.clustering.NGramClusterer$BlockEvaluator.call(NGramClusterer.java:51)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	... 3 more
Caused by: com.google.refine.clustering.knn.DistanceCalculationFailedException: Computing the clusters failed, The distance function returned a non-numeric value for inputs THALIE IMAGES and THALIE IMAGES - TTMT
	at com.google.refine.clustering.knn.UserDefinedDistance.compute(UserDefinedDistance.java:73)
	at com.google.refine.clustering.knn.kNNClusterer$BlockingClusteringRowVisitor$DistanceWrapper.d(kNNClusterer.java:193)
	... 7 more
```

I added the custom exception and caught it in the `ComputeClustersCommand`, but the stack trace is still unclear due to the `ExecutionException` in `/usr/lib/jvm/java-1.11.0-openjdk-amd64!/java.base/java/util/concurrent/FutureTask.class`. This issue also explains why entering "foo" doesn't return an error.

zyadtaha commented:
I reverted the last commit because we need changes in the Vicino library to enable proper error-reporting (like showing "Computing the clusters failed", "The distance function returned a non-numeric value for inputs `"hello"` and `"world"`")

The problem is that the `FutureTask.class` throws an `ExecutionException`, but when it's caught in `NGramClusterer.class`, it doesn't rethrow the exception. This means the error doesn't reach the `ComputeClustersCommand.class` or notify the user, Instead, it just prints the stack trace without alerting the user that something went wrong, so the frontend doesn't get an error response. As a result, we can't decide whether to display an error message to the user or not.

This block of code is from `NGramClusterer.class` in the Vicino library:
```
       try {
            List<Future<Map<Serializable, Set<Serializable>>>> futures = executor.invokeAll(evaluators);
            var18 = futures.iterator();

            while(var18.hasNext()) {
                Future<Map<Serializable, Set<Serializable>>> future = (Future)var18.next();
                cluster_maps.add((Map)future.get());
            }
        } catch (InterruptedException var14) {
            var14.printStackTrace();
        } catch (ExecutionException var15) {
            var15.printStackTrace();
        }
```

zyadtaha commented:
> I think this is good to merge!

Actually, I'm still addressing the issue where the distance clustering is returning incorrect clusters, even in the initial preview.

I've noticed that in #6793, when using the custom Levenshtein function in the main clustering dialog, the number of clusters returned differs from the number obtained with the pre-defined Levenshtein function. This indicates that the problem likely isn't related to the threading of the clustering, as that would also impact the pre-defined Levenshtein function, right?

zyadtaha commented:
Hi @wetneb, any idea why this might be happening?
```
  1 failing

  1) cypress/e2e/create-project/preview_project.cy.js
       Ensures the working of Start-Over Button:
     TypeError: The following error originated from your application code, not from Cypress.
```

wetneb commented:
It's probably just a flaky test, looks like #6825 might fix it.

tfmorris commented:
The Start Over test is one of the flaky ones that I fixed and you can pick up the fix by rebasing onto `master`. If you suspect a flaky test, you can also request that the CI job be rerun to see if that fixes things.
